cpp/src/delayed_submission.cpp

```diff
@@ -34,92 +76,27 @@ bool DelayedSubmissionCollection::isDelayedRequestSubmissionEnabled() const
 
 void DelayedSubmissionCollection::processPre()
 {
-  decltype(_requests) requestsToProcess;
-  {
-    std::lock_guard<std::mutex> lock(_mutex);
-    // Move _requests to a local copy in order to to hold the lock for as
-    // short as possible
-    requestsToProcess = std::move(_requests);
-  }
-  if (requestsToProcess.size() > 0) {
-    ucxx_trace_req("Submitting %lu requests", requestsToProcess.size());
-    for (auto& pair : requestsToProcess) {
-      auto& req      = pair.first;
-      auto& callback = pair.second;
-
-      ucxx_trace_req("Submitting request: %p", req.get());
-
-      if (callback) callback();
-    }
-  }
-  decltype(_genericPre) callbacks;
-  {
-    std::lock_guard<std::mutex> lock(_mutex);
-    // Move _genericPre to a local copy in order to to hold the lock for as
-    // short as possible
-    callbacks = std::move(_genericPre);
-  }
-
-  if (callbacks.size() > 0) {
-    ucxx_trace_req("Submitting %lu generic", callbacks.size());
-
-    for (auto& callback : callbacks) {
-      ucxx_trace_req("Submitting generic");
-
-      if (callback) callback();
-    }
-  }
-}
+  _requests.process();
 
-void DelayedSubmissionCollection::processPost()
-{
-  decltype(_genericPost) callbacks;
-  {
-    std::lock_guard<std::mutex> lock(_mutex);
-    // Move _genericPost to a local copy in order to to hold the lock for as
-    // short as possible
-    callbacks = std::move(_genericPost);
-  }
-
-  if (callbacks.size() > 0) {
-    ucxx_trace_req("Submitting %lu generic", callbacks.size());
-
-    for (auto& callback : callbacks) {
-      ucxx_trace_req("Submitting generic");
-
-      if (callback) callback();
-    }
-  }
+  _genericPre.process();
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**thought (non-blocking):** ‏ While I was reading this I was worried that these two objects were processed in the wrong order, but then I went back over the code and remembered that this is correct. The `genericPre` callbacks are to pass information between threads that the requests have been processed IIRC. Perhaps this is worth documenting here as a comment (or in the docstring?)

Reply:
That's a good idea, [done](https://github.com/rapidsai/ucxx/pull/89/commits/a679e2fe37632a49fce2f2e91215e79b359394e3).